### PR TITLE
optimize topk for greedysearch

### PR DESCRIPTION
**Pull request by yufenglee**
Optimize top 1 computation in greedysearch.
For vocabulary size 50k on A100,
- batch size 1: from 220us to 10.4us.
- batch size 4, from 230us to 11.5us.
For generation of 50 tokens for example, it saves 50*0.2ms = 10ms.

### CODE REVIEW

File: onnxruntime/contrib_ops/cpu/transformers/generate_impl_base.h [hunk 1]
```diff
@@ -33,6 +33,40 @@ gsl::span<T> AllocateBuffer(AllocatorPtr allocator,
...
+    gsl::span<ElementType>& stage_1_scores,  // shape (batch_size, parts_of_vocab)
+    gsl::span<int32_t>& stage_1_tokens,      // shape (batch_size, parts_of_vocab)
+    gsl::span<ElementType>& output_scores,   // shape (batch_size)
+    gsl::span<int32_t>& output_tokens        // shape (batch_size)
+) {
+  constexpr size_t kMaxPartsPerVocab = 128;
+  const size_t stage_1_element_size = kMaxPartsPerVocab * batch_size;
+  const size_t output_element_size = batch_size;
+
+  // Note: use float to allocate buffer for temporary value buffer to avoid unalignment
```
Comment (tianleiwu) [inline onnxruntime/contrib_ops/cpu/transformers/generate_impl_base.h:50]: If Alloc api will get unalignment memory, we shall fix Alloc itself...

File: onnxruntime/contrib_ops/cuda/transformers/greedy_search_top_one.cu [hunk 2]
```diff
@@ -0,0 +1,169 @@
...
+    T* tmp_values,
+    int32_t* tmp_tokens,
+    T* output_values,
+    int32_t* output_tokens,
+    cudaStream_t stream) {
+  constexpr int kThreadBlockSize = GridDim::maxThreadsPerBlock;
+
+  int voc_parts = 4;
+  if (batch_size < 256) {
+    voc_parts = (240 + batch_size - 1) / batch_size;
```
Comment (tianleiwu) [inline onnxruntime/contrib_ops/cuda/transformers/greedy_search_top_one.cu:125]: why 240 is used here instead of 256.

**Review by tianleiwu — APPROVED** [review]

File: onnxruntime/contrib_ops/cuda/transformers/greedy_search_top_one.cu (hunk 2 above)
Comment (tianleiwu) [reply onnxruntime/contrib_ops/cuda/transformers/greedy_search_top_one.cu:125]: I see. Shall we query the SM count and update the parameter accordingly? For example, A100 has 108 SM, then 240 is not optimal here.

File: onnxruntime/contrib_ops/cpu/transformers/generate_impl_base.h (hunk 1 above)
Comment (yufenglee) [reply onnxruntime/contrib_ops/cpu/transformers/generate_impl_base.h:50]: It is not caused by Alloc. One continuous buffer is allocated for stage1_score, stage1_token, stage2_score, stage2_token. If stage_1_element_size is odd and type is fp16, then buffer size for stage1_score may not a multiply of 4. Howerver, stage1_token (int32_t) requires alignment of 4. For simplicity, I just use float to allocate the temp buffer as it is small.

File: onnxruntime/contrib_ops/cuda/transformers/greedy_search_top_one.cu (hunk 2 above)
Comment (yufenglee) [reply onnxruntime/contrib_ops/cuda/transformers/greedy_search_top_one.cu:125]: Will do it in following PR.

**Pull request merged**